sklearn/utils/_array_api.py

```diff
@@ -463,14 +478,14 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
         return float(numpy.average(sample_score_np, weights=sample_weight_np))
 
     if not xp.isdtype(sample_score.dtype, "real floating"):
-        # We move to cpu device ahead of time since certain devices may not support
-        # float64, but we want the same precision for all devices and namespaces.
-        sample_score = xp.astype(xp.asarray(sample_score, device="cpu"), xp.float64)
+        sample_score = xp.astype(
+            sample_score, max_precision_float_dtype(xp, device(sample_score))
+        )
```

Comment:
I prefer the previous code because it guarantees that the precision level is the same for all backends.

Reply:
What is the original reason for changing the dtype of `sample_score`? The way I read the `if` statement is that "if this is a int, then make it a float". Is this the intended way to read this?

Reply:
"not" is not a mistake, because in fact the array api does not specify how to deal with unsafe promotion rules, so we need to explicitly apply the unsafe casting before running operations.